_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+  time = "0.1.35"
+  tz = "0.2.1"
++ nix = "0.5.0"
+```
+
+As before, when I run `cargo build`, Cargo *conservatively* adds `nix` and its dependencies:
+
+```
+$ cargo build
+    Updating registry `https://github.com/rust-lang/crates.io-index`
+ Downloading nix v0.5.0
+ Downloading bitflags v0.4.0
+   Compiling bitflags v0.4.0
+   Compiling nix v0.5.0
+   Compiling datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+But if we look a little closer, we'll notice that `nix` has a dependency on
+`bitflags` **and** `libc`. It now **shares** the dependency on `libc` with the
+`date` package.
+
+If my `datetime` crate gets `libc` types from `time` and hands them off to
+`nix`, they better be the same `libc` or my program won't compile (and we
+wouldn't want it to!)
+
+Today, Cargo will automatically share dependencies between crates if they depend
+on the same **major version** (or minor version before 1.0), since Rust uses
+[semantic versioning](http://semver.org/). This means that if `nix` and `datetime`
+both depend on some version of `libc 0.2.x`, they will get the same version. In
+this case, they do, and the program compiles.
+
+While this policy works well (and in fact is the same policy that system package
+managers use), it doesn't always do exactly what people expect, especially when
+it comes to coordinating a major version bump across the ecosystem. (In many
+cases, it would be preferable for Cargo to hard-error than assume that a
+dependency on `0.2.x` is simply unrelated to another dependency on `0.3.x`.)
+
+This problem is especially pernicious when multiple major versions of the same
+package expose global symbols (using `#[no_mangle]` for example, or by including
+other statically linked C libraries).
+
+We have some thoughts on ways to improve Cargo to handle these cases better,
+including the ability for a package to more explicitly express when a dependency
+is used purely internally and is not shared through its public interface. Those
+packages could be more readily duplicated if needed, while dependencies that are
+used in a package's public interface must not be.
+
+You should expect to see more on this topic in the months ahead.
+
+## Workflow
+
+As we've seen, **Cargo is not just a dependency manager, but Rust's primary
+workflow tool**.
+
+This allows Rust to have a rich ecosystem of interconnected dependencies,
+eliminating the need for applications to manually manage large dependency
+graphs. Applications can benefit from a vibrant ecosystem of small packages that
+do one thing and do it well, and let Cargo handle the heavy lifting of keeping
+everything up to date and compiling correctly.
+
+Even a small crate like the `datetime` example we built has a few dependencies
+on small, targeted crates, and each of those crates has some dependencies of its
+own.
+
+By defining shared, well-known workflows, like "build", "test", "bench", "run",
+and "docs", Cargo provides Rust programmers with a way to think about what
```

Comment:
Oh, also, `cargo doc` is quite neat in that it compiles your own code as well as all your dependencies.